javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,303 @@
+    ) {
+      handleHolder.setHandle("writeHandle", mock())
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameNotInSpec_throwsNoElemException() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    assertFailsWith(
+      NoSuchElementException::class,
+      "Particle TestParticle does not have a handle with name handleButNotEntities."
+    ) {
+      handleHolder.setHandle("handleButNotEntities", mock())
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameInSpec_success() {
+    class DummyHandle : HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("writeHandle" to mock())
+    ) {
+      val writeHandle: Handle by handles
+    }
+
+    val handleHolder = DummyHandle()
+    val mock = mock<Handle>()
+
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.writeHandle).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun detach_allHandles_getUnregistered() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.detach()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getUnregistered() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getClosed() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).close()
+    verify(writeMock, times(1)).close()
+  }
+
+  @Test
+  fun reset_allHandles_allElementsRemoved() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    handleHolder.setHandle("readHandle", mock())
+    handleHolder.setHandle("writeHandle", mock())
+    assertThat(handleHolder.handles).isNotEmpty()
+
+    handleHolder.reset()
```

Comment:
what happens / should happen if `reset` is called without `setHandle` being called before? do we have a test case that covers it (sorry, if i missed it)

Reply:
That's a good test case, I'll write it up now. 

The idea I have for the test: before and after calling `reset`, the `handles` map should be empty.